driver-core/src/main/com/mongodb/internal/async/function/RetryState.java

```diff
@@ -148,19 +181,34 @@ private void doAdvanceOrThrow(final Throwable attemptException,
         if (onlyRuntimeExceptions) {
             assertTrue(isRuntime(attemptException));
         }
-        assertTrue(!isFirstAttempt() || exception == null);
-        Throwable newlyChosenException = transformException(exception, attemptException, onlyRuntimeExceptions, exceptionTransformer);
+        assertTrue(!isFirstAttempt() || previouslyChosenException == null);
+
+        Throwable newlyChosenException = transformException(previouslyChosenException, attemptException, onlyRuntimeExceptions, exceptionTransformer);
+
+        previouslyChosenException = newlyChosenException;
         if (isLastAttempt()) {
-            exception = newlyChosenException;
-            throw exception;
+            /*
+             * Ensure that loopState.isLastIteration() is not true. It returns true when markAsLastIteration() has been called, which
+             * might occur in the following cases:
+             * - All operations in a bulk write successfully complete without transient retryable errors, but some are not acknowledged by
+             *   the server.
+             * - In asynchronous code, breakAndCompleteIf() is used to check if the bulk is ordered, and the operation fails with a
+             *   non-retryable error.
+             * - breakAndCompleteIfRetryAnd() is used to check if the operation is executed in a transaction, and a retry attempt is made.
+             *
+             * In each case, these situations should not be treated as timeout exceptions but rather as exceptions on the server.
+             */
+            if (hasTimeoutMs() && !loopState.isLastIteration()) {
+                throw new MongoOperationTimeoutException("MongoDB operation timed out during a retry attempt", previouslyChosenException);
```

Comment:
In the CSOT [spec](https://github.com/mongodb/specifications/blob/b3e06831a09ae72d94da79f428d8941ec7221f7f/source/client-side-operations-timeout/client-side-operations-timeout.rst#retryability), it’s mentioned to throw a timeout error if the timeout is expired. However, in the retryable read spec, specifically in section [3c. ](https://github.com/mongodb/specifications/blob/master/source/retryable-reads/retryable-reads.rst#id24) where it discusses what to do if a retry attempt fails, the guidance seems a bit different. It suggests throwing a retryable error if the timeout is expired or not set, implying the throwing of the previous error instead of a timeout exception.

The pseudo-code also follows this pattern:
```
else if (isExpired(timeoutMS)) {
      /* CSOT is enabled and the operation has timed out. */
      throw previousError;
    }
```
To comply with both specifications `MongoOperationTimeoutException` is thrown with a previous error as an underlying cause.